third_party/xla_client/thread_pool.cc

```diff
@@ -1,18 +1,180 @@
 #include "tensorflow/compiler/xla/xla_client/thread_pool.h"
 
-#include <thread>
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <mutex>
+
+#include "tensorflow/compiler/xla/xla_client/metrics.h"
+#include "tensorflow/compiler/xla/xla_client/tf_logging.h"
 
 namespace xla {
 namespace env {
+namespace {
+
+class ThreadPool {
+ public:
+  explicit ThreadPool(size_t num_threads) {
+    threads_.reserve(num_threads);
+    for (size_t i = 0; i < num_threads; ++i) {
+      threads_.emplace_back([this]() { Worker(); });
+    }
+  }
+
+  ~ThreadPool() {
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      done_ = true;
+      cv_.notify_all();
+    }
+    for (auto& thread : threads_) {
+      thread.join();
+    }
+  }
+
+  void Schedule(std::function<void()> closure) {
+    // If we have more work scheduled than waiting worker threads, just schedule
+    // it on a separate thread. This prevents tricky thread-pool-size-deadlocks
+    // caused by an undersized thread pool and closures that end up doing sync
+    // waits on the pool threads.
+    bool scheduled = false;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (work_.size() < waiting_) {
+        work_.emplace_back(std::move(closure));
+        scheduled = true;
+      }
+    }
+    if (scheduled) {
+      cv_.notify_one();
+    } else {
+      ScheduleOnThread(std::move(closure));
```

Comment:
Shouldn't we warn on undersized pool? This shouldn't really happen for "good" configurations.

Reply:
It is hard to size a thread pool where the closures that run in them, might block (say, ClientSession::Run()).
The idea is that for normal load, we run on existing threads and avoid create/destroy cycles.
But if we get more load, we offload on new threads.